frontend/test/resolution/testCompilerError.cpp

```diff
@@ -0,0 +1,190 @@
+  guard.realizeErrors();
+}
+
+static void testDepth2() {
+  auto ctx = buildStdContext();
+  ErrorGuard guard(ctx);
+  std::string program =
+    R"""(
+    proc foo(x) {
+      compilerError("some error", 2);
+      return x;
+    }
+    proc bar(x) do return foo(x);
+    var x = bar(42);
+    )""";
+
+  resolveTypesOfVariables(ctx, program, {"x"});
+  ensureErrorInErrorsModule(ctx, guard.errors(), ErrorType::UserDiagnosticEncounterError);
+  ensureErrorOnLine(ctx, guard.errors(), ErrorType::UserDiagnosticEmitError, 7, "some error");
+  guard.realizeErrors();
+}
+
+static void testDepth3() {
+  auto ctx = buildStdContext();
+  ErrorGuard guard(ctx);
+  std::string program =
+    R"""(
+    proc foo(x) {
+      compilerError("some error", 3);
+      return x;
+    }
+    proc bar(x) do return foo(x);
+    proc baz(x) do return bar(x);
+    var x = baz(42);
+    )""";
+
+  resolveTypesOfVariables(ctx, program, {"x"});
+  ensureErrorInErrorsModule(ctx, guard.errors(), ErrorType::UserDiagnosticEncounterError);
+  ensureErrorOnLine(ctx, guard.errors(), ErrorType::UserDiagnosticEmitError, 8, "some error");
+  guard.realizeErrors();
+}
+
+static void testTooDeep() {
+  auto ctx = buildStdContext();
+  ErrorGuard guard(ctx);
+  std::string program =
+    R"""(
+    proc foo(x) {
+      compilerError("some error", 1000);
+      return x;
+    }
+    var x = foo(42);
+    )""";
+
+  resolveTypesOfVariables(ctx, program, {"x"});
+  ensureErrorInErrorsModule(ctx, guard.errors(), ErrorType::UserDiagnosticEncounterError);
+  ensureErrorOnLine(ctx, guard.errors(), ErrorType::UserDiagnosticEmitError, 6, "some error");
+  guard.realizeErrors();
+}
+
+static void testVarArgs() {
+  auto ctx = buildStdContext();
+  ErrorGuard guard(ctx);
+  std::string program =
+    R"""(
+    proc foo(x) {
+      compilerError("this ", "is ", "some ", "error ", "message");
+      return x;
+    }
+    var x = foo(42);
+    )""";
+
+  resolveTypesOfVariables(ctx, program, {"x"});
+  ensureErrorInErrorsModule(ctx, guard.errors(), ErrorType::UserDiagnosticEncounterError);
+  ensureErrorOnLine(ctx, guard.errors(), ErrorType::UserDiagnosticEmitError, 6, "this is some error message");
+  guard.realizeErrors();
+}
+
+static void testRunAndTrackErrors() {
+  auto ctx = buildStdContext();
+  ErrorGuard guard(ctx);
+  std::string program =
+    R"""(
+    proc foo() {
+      compilerError("some error");
+      return 42;
+    }
+    )""";
+
+  auto file = UniqueString::get(ctx, "test.chpl");
+  setFileText(ctx, file, program);
+  auto modules = parse(ctx, file, UniqueString());
+  assert(modules.size() == 1);
+  assert(modules[0]->numStmts() == 1);
+
+  auto result = ctx->runAndTrackErrors([&](Context* ctx) {
+    return resolveConcreteFunction(ctx, modules[0]->stmt(0)->id());
+  });
+  assert(!result.ranWithoutErrors());
+  assert(result.errors().size() == 1);
+  ensureErrorInErrorsModule(ctx, result.errors(), ErrorType::UserDiagnosticEncounterError);
+}
+
+int main() {
```

Comment:
I think it would be good to test a case where two different functions call the same instantiation of a function that contains either a warning or an error. And perhaps have those callers have different depths? E.g.,

```chpl
proc foo() { ... compilerWarning(..., 2); }

proc first() { foo(); }
proc second() { foo(); }
proc third() { second(); }

first();
third();
```